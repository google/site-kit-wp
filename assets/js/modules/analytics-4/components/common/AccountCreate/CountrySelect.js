/**
 * CountrySelect component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { Option, Select } from 'googlesitekit-components';
import { useSelect, useDispatch } from 'googlesitekit-data';
import {
	allCountries,
	countriesByCode,
} from '../../../utils/countries-timezones';
import { FORM_ACCOUNT_CREATE } from '../../../datastore/constants';
import { CORE_FORMS } from '../../../../../googlesitekit/datastore/forms/constants';

export default function CountrySelect() {
	const value = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_ACCOUNT_CREATE, 'countryCode' )
	);

	const { setValues } = useDispatch( CORE_FORMS );
	const onEnhancedChange = useCallback(
		( i, item ) => {
			const newCountryCode = item.dataset.value;
			if (
				newCountryCode !== value &&
				countriesByCode[ newCountryCode ]
			) {
				setValues( FORM_ACCOUNT_CREATE, {
					countryCode: newCountryCode,
					timezone:
						// eslint-disable-next-line sitekit/acronym-case
						countriesByCode[ newCountryCode ].defaultTimeZoneId,
				} );
			}
		},
		[ setValues, value ]
	);

	return (
		<Select
			className="googlesitekit-analytics__select-country"
			label={ __( 'Country', 'google-site-kit' ) }
			value={ value }
			onEnhancedChange={ onEnhancedChange }
			enhanced
			outlined
		>
			{ allCountries.map( ( { countryCode, displayName }, i ) => (
				<Option key={ i } value={ countryCode }>
					{ displayName }
				</Option>
			) ) }
		</Select>
	);
}
