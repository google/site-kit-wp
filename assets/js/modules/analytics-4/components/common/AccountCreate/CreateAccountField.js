/**
 * CreateAccountField component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import classnames from 'classnames';

/**
 * Internal dependencies
 */
import { TextField } from 'googlesitekit-components';

export default function CreateAccountField( {
	hasError,
	value,
	setValue,
	name,
	label,
} ) {
	// Ensure field doesn't render until default value is available, fixing a potential render bug.
	if ( value === undefined ) {
		return null;
	}

	return (
		<TextField
			className={ classnames( 'mdc-text-field', {
				'mdc-text-field--error': hasError,
			} ) }
			label={ label }
			name={ name }
			onChange={ ( event ) => {
				setValue( event.target.value, name );
			} }
			value={ value }
			id={ `googlesitekit_analytics_account_create_${ name }` }
			outlined
		/>
	);
}
