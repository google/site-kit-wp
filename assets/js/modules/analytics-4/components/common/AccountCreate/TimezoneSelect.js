/**
 * TimezoneSelect component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { Option, Select } from 'googlesitekit-components';
import { useSelect, useDispatch } from 'googlesitekit-data';
import { timeZonesByCountryCode } from '../../../utils/countries-timezones';
import { FORM_ACCOUNT_CREATE } from '../../../datastore/constants';
import { CORE_FORMS } from '../../../../../googlesitekit/datastore/forms/constants';

export default function TimezoneSelect() {
	const countryCode = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_ACCOUNT_CREATE, 'countryCode' )
	);
	const value = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_ACCOUNT_CREATE, 'timezone' )
	);

	const { setValues } = useDispatch( CORE_FORMS );
	const onEnhancedChange = useCallback(
		( i, item ) => {
			setValues( FORM_ACCOUNT_CREATE, { timezone: item.dataset.value } );
		},
		[ setValues ]
	);

	return (
		<Select
			className="googlesitekit-analytics__select-timezone"
			label={ __( 'Timezone', 'google-site-kit' ) }
			value={ value }
			onEnhancedChange={ onEnhancedChange }
			disabled={ ! countryCode }
			enhanced
			outlined
		>
			{ ( timeZonesByCountryCode[ countryCode ] || [] ).map(
				// eslint-disable-next-line sitekit/acronym-case
				( { timeZoneId, displayName }, i ) => (
					<Option
						key={ i }
						// eslint-disable-next-line sitekit/acronym-case
						value={ timeZoneId }
					>
						{ displayName }
					</Option>
				)
			) }
		</Select>
	);
}
