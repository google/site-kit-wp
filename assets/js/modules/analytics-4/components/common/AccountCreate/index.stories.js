/**
 * AccountCreate Component Stories.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import AccountCreate from '.';
import {
	provideModuleRegistrations,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
} from '../../../../../../../tests/js/utils';
import {
	ACCOUNT_CREATE,
	EDIT_SCOPE,
	MODULES_ANALYTICS_4,
} from '../../../datastore/constants';
import WithRegistrySetup from '../../../../../../../tests/js/WithRegistrySetup';
import * as fixtures from '../../../../analytics-4/datastore/__fixtures__';
import { Cell, Grid, Row } from '../../../../../material-components';

const { accountSummaries } = fixtures;

function Template( args ) {
	return (
		<div className="googlesitekit-layout">
			<div className="googlesitekit-settings-module googlesitekit-settings-module--active googlesitekit-settings-module--analytics">
				<div className="googlesitekit-setup-module">
					<div className="googlesitekit-settings-module__content googlesitekit-settings-module__content--open">
						<Grid>
							<Row>
								<Cell size={ 12 }>
									<AccountCreate { ...args } />
								</Cell>
							</Row>
						</Grid>
					</div>
				</div>
			</div>
		</div>
	);
}

export const Default = Template.bind( {} );
Default.storyName = 'AccountCreate';

export default {
	title: 'Modules/Analytics4/Components/AccountCreate',
	decorators: [
		( Story ) => {
			const setupRegistry = ( registry ) => {
				provideModules( registry, [
					{
						slug: 'analytics-4',
						active: true,
						connected: true,
						owner: { login: 'analytics_4-owner-username' },
					},
				] );
				provideSiteInfo( registry );
				provideModuleRegistrations( registry );
				provideUserAuthentication( registry, {
					grantedScopes: [ EDIT_SCOPE ],
				} );

				registry
					.dispatch( MODULES_ANALYTICS_4 )
					.receiveGetAccountSummaries( accountSummaries );

				registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
					accountID: ACCOUNT_CREATE,
					useSnippet: true,
				} );
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
