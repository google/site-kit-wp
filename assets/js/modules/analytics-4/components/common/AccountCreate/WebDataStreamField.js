/**
 * WebDataStreamField component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import CreateAccountField from './CreateAccountField';
import { useSelect, useDispatch } from 'googlesitekit-data';
import { FORM_ACCOUNT_CREATE } from '../../../datastore/constants';
import { CORE_FORMS } from '../../../../../googlesitekit/datastore/forms/constants';

export default function WebDataStreamField() {
	const value = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_ACCOUNT_CREATE, 'dataStreamName' )
	);
	const { setValues } = useDispatch( CORE_FORMS );

	const setValue = useCallback(
		( dataStreamName ) => {
			setValues( FORM_ACCOUNT_CREATE, { dataStreamName } );
		},
		[ setValues ]
	);

	return (
		<CreateAccountField
			label={ __( 'Web Data Stream', 'google-site-kit' ) }
			value={ value }
			hasError={ ! value }
			setValue={ setValue }
			name="dataStream"
		/>
	);
}
