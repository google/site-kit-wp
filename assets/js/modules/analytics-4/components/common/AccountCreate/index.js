/**
 * AccountCreate component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { useCallback, useState, useEffect } from '@wordpress/element';

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import { useSelect, useDispatch } from 'googlesitekit-data';
import { Button, ProgressBar } from 'googlesitekit-components';
import {
	FORM_ACCOUNT_CREATE,
	EDIT_SCOPE,
	GTM_SCOPE,
	MODULES_ANALYTICS_4,
} from '../../../datastore/constants';
import { CORE_SITE } from '../../../../../googlesitekit/datastore/site/constants';
import { CORE_USER } from '../../../../../googlesitekit/datastore/user/constants';
import { CORE_FORMS } from '../../../../../googlesitekit/datastore/forms/constants';
import { CORE_LOCATION } from '../../../../../googlesitekit/datastore/location/constants';
import { ERROR_CODE_MISSING_REQUIRED_SCOPE } from '../../../../../util/errors';
import { trackEvent } from '../../../../../util';
import { getAccountDefaults as getAccountDefaults } from '../../../utils/account';
import { Cell } from '../../../../../material-components';
import StoreErrorNotices from '../../../../../components/StoreErrorNotices';
import TimezoneSelect from './TimezoneSelect';
import AccountField from './AccountField';
import PropertyField from './PropertyField';
import CountrySelect from './CountrySelect';
import WebDataStreamField from './WebDataStreamField';
import EnhancedMeasurementSwitch from '../EnhancedMeasurementSwitch';
import useViewContext from '../../../../../hooks/useViewContext';
import SetupEnhancedConversionTrackingNotice from '../../../../../components/conversion-tracking/SetupEnhancedConversionTrackingNotice';

export default function AccountCreate() {
	const [ isNavigating, setIsNavigating ] = useState( false );
	const accounts = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getAccountSummaries()
	);

	const hasResolvedAccounts = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
			'getAccountSummaries'
		)
	);
	const accountTicketTermsOfServiceURL = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getAccountTicketTermsOfServiceURL()
	);
	const canSubmitAccountCreate = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).canSubmitAccountCreate()
	);
	const isDoingCreateAccount = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).isDoingCreateAccount()
	);
	const hasEditScope = useSelect( ( select ) =>
		select( CORE_USER ).hasScope( EDIT_SCOPE )
	);
	const hasGTMScope = useSelect( ( select ) =>
		select( CORE_USER ).hasScope( GTM_SCOPE )
	);
	const hasAccountCreateForm = useSelect( ( select ) =>
		select( CORE_FORMS ).hasForm( FORM_ACCOUNT_CREATE )
	);
	const autoSubmit = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_ACCOUNT_CREATE, 'autoSubmit' )
	);
	const siteURL = useSelect( ( select ) =>
		select( CORE_SITE ).getReferenceSiteURL()
	);
	const siteName = useSelect( ( select ) =>
		select( CORE_SITE ).getSiteName()
	);
	const timezone = useSelect( ( select ) =>
		select( CORE_SITE ).getTimezone()
	);

	const viewContext = useViewContext();
	const { setValues } = useDispatch( CORE_FORMS );
	const { navigateTo } = useDispatch( CORE_LOCATION );
	const { createAccount } = useDispatch( MODULES_ANALYTICS_4 );
	const { setPermissionScopeError } = useDispatch( CORE_USER );
	const { setConversionTrackingEnabled, saveConversionTrackingSettings } =
		useDispatch( CORE_SITE );

	const hasRequiredScope = hasEditScope;

	// Redirect if the accountTicketTermsOfServiceURL is set.
	useEffect( () => {
		if ( accountTicketTermsOfServiceURL ) {
			( async () => {
				await API.invalidateCache( 'modules', 'analytics-4' );
				navigateTo( accountTicketTermsOfServiceURL );
			} )();
		}
	}, [ accountTicketTermsOfServiceURL, navigateTo ] );

	// Set form defaults on initial render.
	useEffect( () => {
		// Only set the form if not already present in store.
		// e.g. after a snapshot has been restored.
		if ( ! hasAccountCreateForm ) {
			setValues(
				FORM_ACCOUNT_CREATE,
				getAccountDefaults( {
					siteName,
					siteURL,
					timezone,
				} )
			);
		}
	}, [ hasAccountCreateForm, siteName, siteURL, timezone, setValues ] );

	const handleSubmit = useCallback( async () => {
		const scopes = [];

		if ( ! hasEditScope ) {
			scopes.push( EDIT_SCOPE );
		}
		// The GTM scope should be granted for GTE support, but
		// it is possible for it not to be at this point.
		// This saves an extra OAuth flow and is necessary for the
		// Google tag sync at the end of the post-provisioning flow.
		if ( ! hasGTMScope ) {
			scopes.push( GTM_SCOPE );
		}

		// If scope not granted, trigger scope error right away. These are
		// typically handled automatically based on API responses, but
		// this particular case has some special handling to improve UX.
		if ( scopes.length > 0 ) {
			// When state is restored, auto-submit the request again.
			setValues( FORM_ACCOUNT_CREATE, { autoSubmit: true } );
			setPermissionScopeError( {
				code: ERROR_CODE_MISSING_REQUIRED_SCOPE,
				message: __(
					'Additional permissions are required to create a new Analytics account.',
					'google-site-kit'
				),
				data: {
					status: 403,
					scopes,
					skipModal: true,
				},
			} );
			return;
		}

		setValues( FORM_ACCOUNT_CREATE, { autoSubmit: false } );
		await trackEvent(
			`${ viewContext }_analytics`,
			'create_account',
			'proxy'
		);

		const { error } = await createAccount();
		if ( ! error ) {
			setConversionTrackingEnabled( true );
			await saveConversionTrackingSettings();
			setIsNavigating( true );
		}
	}, [
		hasEditScope,
		hasGTMScope,
		setValues,
		viewContext,
		createAccount,
		setPermissionScopeError,
		setConversionTrackingEnabled,
		saveConversionTrackingSettings,
	] );

	// If the user ends up back on this component with the required scope granted,
	// and already submitted the form, trigger the submit again.
	useEffect( () => {
		if ( hasRequiredScope && autoSubmit ) {
			handleSubmit();
		}
	}, [ hasRequiredScope, autoSubmit, handleSubmit ] );

	// If the user clicks "Back", rollback settings to restore saved values, if any.
	const { rollbackSettings } = useDispatch( MODULES_ANALYTICS_4 );
	const handleBack = useCallback(
		() => rollbackSettings(),
		[ rollbackSettings ]
	);

	if (
		isDoingCreateAccount ||
		isNavigating ||
		! hasResolvedAccounts ||
		hasRequiredScope === undefined
	) {
		return <ProgressBar />;
	}

	return (
		<div>
			<StoreErrorNotices
				moduleSlug="analytics-4"
				storeName={ MODULES_ANALYTICS_4 }
			/>

			<h3 className="googlesitekit-heading-4">
				{ __( 'Create your Analytics account', 'google-site-kit' ) }
			</h3>

			<p>
				{ __(
					'Weâ€™ve pre-filled the required information for your new account. Confirm or edit any details:',
					'google-site-kit'
				) }
			</p>

			<div className="googlesitekit-setup-module__inputs">
				<Cell size={ 6 }>
					<AccountField />
				</Cell>
				<Cell size={ 6 }>
					<PropertyField />
				</Cell>
				<Cell size={ 6 }>
					<WebDataStreamField />
				</Cell>
			</div>

			<div className="googlesitekit-setup-module__inputs">
				<CountrySelect />

				<TimezoneSelect />
			</div>

			<div className="googlesitekit-setup-module__inputs">
				<EnhancedMeasurementSwitch
					formName={ FORM_ACCOUNT_CREATE }
					className="googlesitekit-margin-bottom-0"
				/>

				<SetupEnhancedConversionTrackingNotice
					className="googlesitekit-margin-top-0"
					message={ __(
						'To track how visitors interact with your site, Site Kit will enable enhanced conversion tracking. You can always disable it in settings.',
						'google-site-kit'
					) }
				/>
			</div>

			<p>
				{ hasRequiredScope && (
					<span>
						{ __(
							'You will be redirected to Google Analytics to accept the terms of service.',
							'google-site-kit'
						) }
					</span>
				) }
				{ ! hasRequiredScope && (
					<span>
						{ __(
							'You will need to give Site Kit permission to create an Analytics account on your behalf and also accept the Google Analytics terms of service.',
							'google-site-kit'
						) }
					</span>
				) }
			</p>

			<div className="googlesitekit-setup-module__action">
				<Button
					disabled={ ! canSubmitAccountCreate }
					onClick={ handleSubmit }
				>
					{ __( 'Create Account', 'google-site-kit' ) }
				</Button>

				{ accounts && !! accounts.length && (
					<Button
						tertiary
						className="googlesitekit-setup-module__sub-action"
						onClick={ handleBack }
					>
						{ __( 'Back', 'google-site-kit' ) }
					</Button>
				) }
			</div>
		</div>
	);
}
