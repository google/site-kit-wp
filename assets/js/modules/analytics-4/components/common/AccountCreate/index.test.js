/**
 * AccountCreate component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	createTestRegistry,
	fireEvent,
	muteFetch,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
	provideUserInfo,
	render,
} from '../../../../../../../tests/js/test-utils';
import { mockLocation } from '../../../../../../../tests/js/mock-browser-utils';
import { mockUseInstanceID } from '../../../../../../../tests/js/mock-use-instance-id';

import {
	EDIT_SCOPE,
	GTM_SCOPE,
	MODULES_ANALYTICS_4,
	PROVISIONING_SCOPE,
} from '../../../datastore/constants';
import { createCacheKey } from '../../../../../googlesitekit/api';
import { getKeys, setItem } from '../../../../../googlesitekit/api/cache';
import AccountCreate from '.';
import { CORE_SITE } from '../../../../../googlesitekit/datastore/site/constants';

const REGEX_REST_CONVERSION_TRACKING_SETTINGS = new RegExp(
	'^/google-site-kit/v1/core/site/data/conversion-tracking'
);

describe( 'AccountCreate', () => {
	mockLocation();
	// Provide our own version of the `useInstanceId()` hook to avoid instability in test snapshots.
	mockUseInstanceID();

	let registry;

	beforeEach( () => {
		registry = createTestRegistry();
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideSiteInfo( registry );
		provideUserInfo( registry );
		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE, PROVISIONING_SCOPE, GTM_SCOPE ],
		} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries( {
			accountSummaries: [],
			nextPageToken: null,
		} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.finishResolution( 'getAccountSummaries', [] );

		// Enable Enhanced Conversion Tracking by default to avoid adding
		// the notice in existing cases.
		registry.dispatch( CORE_SITE ).receiveGetConversionTrackingSettings( {
			enabled: true,
		} );
	} );

	it( 'renders correctly in a loading state', async () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.invalidateResolution( 'getAccountSummaries', [] );

		const { container, getByRole, waitForRegistry } = render(
			<AccountCreate />,
			{
				registry,
			}
		);

		expect( container ).toMatchSnapshot();

		expect( getByRole( 'progressbar' ) ).toBeInTheDocument();

		// Wait for resolvers to finish to avoid an unhandled React state update.
		await waitForRegistry();
	} );

	it( 'renders correctly when data has loaded', async () => {
		const { container, getByRole, waitForRegistry } = render(
			<AccountCreate />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect(
			getByRole( 'button', { name: 'Create Account' } )
		).toBeInTheDocument();
	} );

	describe( 'when clicking on Create Account', () => {
		const accountTicketID = 'abc123';

		let getByRole, waitForRegistry;

		beforeEach( () => {
			( { getByRole, waitForRegistry } = render( <AccountCreate />, {
				registry,
			} ) );

			fetchMock.post(
				new RegExp(
					'^/google-site-kit/v1/modules/analytics-4/data/create-account-ticket'
				),
				{
					// eslint-disable-next-line sitekit/acronym-case
					body: { accountTicketId: accountTicketID },
					status: 200,
				}
			);
			muteFetch( REGEX_REST_CONVERSION_TRACKING_SETTINGS );
		} );

		it( 'should invalidate the module cache', async () => {
			await setItem(
				createCacheKey(
					'modules',
					'analytics-4',
					'analytics-datapoint'
				),
				'analytics-value'
			);

			const searchConsoleItemCacheKey = createCacheKey(
				'modules',
				'search-console',
				'search-console-datapoint'
			);

			await setItem( searchConsoleItemCacheKey, 'search-console-value' );

			await expect( getKeys() ).resolves.toHaveLength( 2 );

			fireEvent.click(
				getByRole( 'button', { name: 'Create Account' } )
			);

			await waitForRegistry();

			// Verify the cache is cleared for the `analytics-4` module.
			const cacheKeys = await getKeys();
			expect( cacheKeys ).toHaveLength( 1 );
			expect( cacheKeys[ 0 ] ).toContain( searchConsoleItemCacheKey );
		} );

		it( 'should navigate to the Google Analytics Terms of Service', async () => {
			fireEvent.click(
				getByRole( 'button', { name: 'Create Account' } )
			);

			await waitForRegistry();

			const accountTicketTermsOfServiceURL = registry
				.select( MODULES_ANALYTICS_4 )
				.getAccountTicketTermsOfServiceURL();

			// Verify that the URL has been navigated to.
			expect( global.location.assign ).toHaveBeenCalledWith(
				accountTicketTermsOfServiceURL
			);

			// Sanity check that the URL is correct.
			expect( accountTicketTermsOfServiceURL ).toMatch(
				new RegExp(
					`analytics.google.com.*termsofservice.*${ accountTicketID }`
				)
			);
		} );

		it( 'should enable conversion tracking', async () => {
			fireEvent.click(
				getByRole( 'button', { name: 'Create Account' } )
			);

			await waitForRegistry();

			expect( fetchMock ).toHaveFetched(
				REGEX_REST_CONVERSION_TRACKING_SETTINGS
			);
		} );
	} );
} );
