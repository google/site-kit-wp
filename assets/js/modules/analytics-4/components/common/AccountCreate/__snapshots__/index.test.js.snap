// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCreate renders correctly in a loading state 1`] = `
<div>
  <div
    class="mdc-linear-progress mdc-linear-progress--indeterminate"
    role="progressbar"
  >
    <div
      class="mdc-linear-progress__buffering-dots"
    />
    <div
      class="mdc-linear-progress__buffer"
    />
    <div
      class="mdc-linear-progress__bar mdc-linear-progress__primary-bar"
    >
      <span
        class="mdc-linear-progress__bar-inner"
      />
    </div>
    <div
      class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"
    >
      <span
        class="mdc-linear-progress__bar-inner"
      />
    </div>
  </div>
</div>
`;

exports[`AccountCreate renders correctly when data has loaded 1`] = `
<div>
  <div>
    <h3
      class="googlesitekit-heading-4"
    >
      Create your Analytics account
    </h3>
    <p>
      We’ve pre-filled the required information for your new account. Confirm or edit any details:
    </p>
    <div
      class="googlesitekit-setup-module__inputs"
    >
      <div
        class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6"
      >
        <div
          class="mdc-text-field mdc-text-field mdc-text-field--outlined"
          name="account"
        >
          <input
            class="mdc-text-field__input"
            id="googlesitekit_analytics_account_create_account"
            value="My Site Name"
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <label
                class="mdc-floating-label mdc-floating-label--float-above"
                for="googlesitekit_analytics_account_create_account"
                style=""
              >
                Account
              </label>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </div>
      </div>
      <div
        class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6"
      >
        <div
          class="mdc-text-field mdc-text-field mdc-text-field--outlined"
          name="property"
        >
          <input
            class="mdc-text-field__input"
            id="googlesitekit_analytics_account_create_property"
            value="example.com"
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <label
                class="mdc-floating-label mdc-floating-label--float-above"
                for="googlesitekit_analytics_account_create_property"
                style=""
              >
                Property
              </label>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </div>
      </div>
      <div
        class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6"
      >
        <div
          class="mdc-text-field mdc-text-field mdc-text-field--outlined"
          name="dataStream"
        >
          <input
            class="mdc-text-field__input"
            id="googlesitekit_analytics_account_create_dataStream"
            value="example.com"
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <label
                class="mdc-floating-label mdc-floating-label--float-above"
                for="googlesitekit_analytics_account_create_dataStream"
                style=""
              >
                Web Data Stream
              </label>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-setup-module__inputs"
    >
      <div
        class="mdc-select googlesitekit-analytics__select-country mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="US"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          United States
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-0"
              style=""
            >
              Country
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
      <div
        class="mdc-select googlesitekit-analytics__select-timezone mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="America/Detroit"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          (GMT-04:00) Detroit Time
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-1"
              style=""
            >
              Timezone
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-setup-module__inputs"
    >
      <div
        class="googlesitekit-analytics-enable-enhanced-measurement googlesitekit-margin-bottom-0"
      >
        <div
          aria-checked="true"
          class="mdc-switch mdc-switch--checked"
          role="switch"
          tabindex="0"
        >
          <div
            class="mdc-switch__track"
          >
             
          </div>
          <div
            class="mdc-switch__thumb-underlay"
          >
            <div
              class="mdc-switch__thumb"
            >
              <input
                class="mdc-switch__native-control"
                id="googlesitekit-switch-0"
                readonly=""
                role="switch"
                type="checkbox"
              />
            </div>
          </div>
        </div>
        <label
          class=""
          for="googlesitekit-switch-0"
        >
          Enable enhanced measurement
        </label>
        <p
          class="googlesitekit-module-settings-group__helper-text"
        >
          This allows you to measure interactions with your content (e.g. file downloads, form completions, video views). 
          <a
            aria-label="Learn more (opens in a new tab)"
            class="googlesitekit-cta-link"
            href="https://support.google.com/analytics/answer/9216061?hl=en-US"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="googlesitekit-cta-link__contents"
            >
              Learn more
            </span>
            <span
              class="googlesitekit-icon-wrapper"
              style="margin-left: 5px;"
            >
              <svg />
            </span>
          </a>
        </p>
      </div>
    </div>
    <p>
      <span>
        You will be redirected to Google Analytics to accept the terms of service.
      </span>
    </p>
    <div
      class="googlesitekit-setup-module__action"
    >
      <button
        class="mdc-button mdc-button--raised"
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          Create Account
        </span>
      </button>
    </div>
  </div>
</div>
`;
